lib/ash/filter/filter.ex

```diff
@@ -3664,6 +3671,25 @@ defmodule Ash.Filter do
     end
   end
 
+  def do_hydrate_refs(
```

Comment:
hydrating refs can happen before the context we need is present. I think here needs to leave the `%Ash.Query.UpsertConflict{}` in place, and instead this logic should go in `filter/runtime.ex`

Reply:
@zachdaniel Yes, it already hydrates without the context in `Ash.Changeset.filter/2`.

This is why there's two function clauses defined.

My idea was the following:

* If the context is not present, leave the expression as is
* If the context is present, the adapter is setting the relevant info and we can hydrate it directly
* If the filter is run when the adapter didn't override the expression (as seen in the postgres PR) and does not provide the context (as with `ETS`), the call will land in runtime
* When reaching runtime, it will error that the datasource does not support the expression.

Reply:
I'm taking this back. When we use `filter_matches`, it works differently.

Reply:
I'm taking back the taking back :laughing: I just did it wrong.